File: code/data_storage/java/data/src/java/com/ggeoffre/database/PostgresData.java
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2025-2026 ggeoffre, LLC

package com.geoffrey.database;

import com.geoffrey.database.SensorDataJsonHelper;
import java.sql.*;
import java.util.*;
import org.json.*;

public class PostgresData {

    // Configuration constants
    private static final String HOST = "192.168.1.60";
    private static final int PORT = 5432;
    private static final String USER = "postgres";
    private static final String PASSWORD = "";
    private static final String DATABASE = "sensor_data_db";
    private static final String TABLE = "sensor_data";

    public static void processSensorData() {
        System.out.println("POSTGRES");
        System.out.println("########");

        // Step 1: Connect to the default "postgres" database to create the target database if needed
        String urlPostgres = String.format(
            "jdbc:postgresql://%s:%d/postgres",
            HOST,
            PORT
        );
        try (
            Connection connection = DriverManager.getConnection(
                urlPostgres,
                USER,
                PASSWORD
            )
        ) {
            System.out.println("Connected to Postgres (postgres database)");

            // Check if database exists, create if not
            try (Statement stmt = connection.createStatement()) {
                String checkDbSql = String.format(
                    "SELECT 1 FROM pg_database WHERE datname = '%s'",
                    DATABASE
                );
                ResultSet rs = stmt.executeQuery(checkDbSql);
                boolean dbExists = rs.next();
                rs.close();

                if (!dbExists) {
                    String createDbSql = String.format(
                        "CREATE DATABASE %s;",
                        DATABASE
                    );
                    stmt.executeUpdate(createDbSql);
                    System.out.println("Database created: " + DATABASE);
                } else {
                    System.out.println("Database already exists: " + DATABASE);
                }
            } catch (SQLException e) {
                System.err.println(
                    "Error creating database: " + e.getMessage()
                );
                e.printStackTrace();
                return;
            }
        } catch (SQLException e) {
            System.err.println(
                "Failed to connect to the Postgres server (postgres database)."
            );
            e.printStackTrace();
            return;
        }

        // Step 2: Connect to the target database for all further operations
        String urlTargetDb = String.format(
            "jdbc:postgresql://%s:%d/%s",
            HOST,
            PORT,
            DATABASE
        );
        try (
            Connection connection = DriverManager.getConnection(
                urlTargetDb,
                USER,
                PASSWORD
            )
        ) {
            System.out.println("Connected to Postgres (" + DATABASE + ")");

            // Create table if not exists
            try (Statement stmt = connection.createStatement()) {
                String createTableSql = String.format(
                    "CREATE TABLE IF NOT EXISTS %s (" +
                        "recorded BIGINT NOT NULL, " +
                        "location VARCHAR(255) NOT NULL, " +
                        "sensor VARCHAR(255) NOT NULL, " +
                        "measurement VARCHAR(255) NOT NULL, " +
                        "units VARCHAR(255) NOT NULL, " +
                        "value NUMERIC(10, 2) NOT NULL" +
                        ");",
                    TABLE
                );
                stmt.executeUpdate(createTableSql);
                System.out.println("Table ensured: " + TABLE);
            } catch (SQLException e) {
                System.err.println("Error creating table: " + e.getMessage());
                e.printStackTrace();
                return;
            }

            // Store sensor data
            String insertSql =
                "INSERT INTO sensor_data (recorded, location, sensor, measurement, units, value) VALUES (?, ?, ?, ?, ?, ?)";
            try {
                JSONObject jsonObj =
                    SensorDataJsonHelper.getSensorDataJSONObject();
                System.out.println("SensorData marshalled to JSON");

                // Validate or generate 'recorded' field
                long recorded = jsonObj.has("recorded")
                    ? jsonObj.getLong("recorded")
                    : System.currentTimeMillis() / 1000;

                try (
                    PreparedStatement pstmt = connection.prepareStatement(
                        insertSql
                    )
                ) {
                    pstmt.setLong(1, recorded);
                    pstmt.setString(2, jsonObj.getString("location"));
                    pstmt.setString(3, jsonObj.getString("sensor"));
                    pstmt.setString(4, jsonObj.getString("measurement"));
                    pstmt.setString(5, jsonObj.getString("units"));
                    pstmt.setBigDecimal(6, jsonObj.getBigDecimal("value"));
                    pstmt.executeUpdate();
                }
                System.out.println("SensorData stored");
            } catch (SQLException | JSONException e) {
                System.err.println("Failed to store sensor data.");
                e.printStackTrace();
            }

            // Retrieve sensor data
            String selectSql =
                "SELECT recorded, location, sensor, measurement, units, value FROM sensor_data";
            List<String> jsonList = new ArrayList<>();
            try (
                Statement stmt = connection.createStatement();
                ResultSet rs = stmt.executeQuery(selectSql)
            ) {
                while (rs.next()) {
                    JSONObject jsonObj = new JSONObject();
                    jsonObj.put("recorded", rs.getLong("recorded"));
                    jsonObj.put("location", rs.getString("location"));
                    jsonObj.put("sensor", rs.getString("sensor"));
                    jsonObj.put("measurement", rs.getString("measurement"));
                    jsonObj.put("units", rs.getString("units"));
                    jsonObj.put("value", rs.getBigDecimal("value"));
                    jsonList.add(jsonObj.toString());
                }
                System.out.println("SensorData retrieved");
            } catch (SQLException | JSONException e) {
                System.err.println("Failed to retrieve sensor data.");
                e.printStackTrace();
            }

            // Print as CSV
            try {
                System.out.print(
                    SensorDataJsonHelper.jsonArrayToCsv(
                        jsonList.toArray(new String[0])
                    )
                );
            } catch (Exception e) {
                System.err.println("Failed to convert JSON to CSV.");
                e.printStackTrace();
            }

            // Purge sensor data (use with caution in production!)
            String deleteSql = "DELETE FROM sensor_data";
            try (Statement stmt = connection.createStatement()) {
                stmt.executeUpdate(deleteSql);
                System.out.println("Sensor Data purged\n");
            } catch (SQLException e) {
                System.err.println("Failed to purge sensor data.");
                e.printStackTrace();
            }
        } catch (SQLException e) {
            System.err.println("Failed to connect to the Postgres database.");
            e.printStackTrace();
        }
    }
}

File: code/data_storage/go/data/database/postgres_data.go
// SPDX-License-Identifier: GPL-3.0-or-later
// Copyright (C) 2025-2026 ggeoffre, LLC

package database

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"strings"

	_ "github.com/jackc/pgx/v5/stdlib"
)

type SensorDataPostgres struct {
	Recorded    int64   `postgres:"recorded"`
	Location    string  `postgres:"location"`
	Sensor      string  `postgres:"sensor"`
	Measurement string  `postgres:"measurement"`
	Units       string  `postgres:"units"`
	Value       float64 `postgres:"value"`
}

// PostgresData struct for Postgres
type PostgresData struct {
	DB *sql.DB
}

// NewPostgresData creates a new PostgresData instance and connects
func NewPostgresData() (*PostgresData, error) {
	// 1. Connect to default 'postgres' database to check/create target DB
	db, err := sql.Open("pgx", "postgres://postgres:postgres@localhost:5432/postgres?sslmode=disable")
	if err != nil {
		return nil, fmt.Errorf("failed to open default database: %w", err)
	}

	databaseName := "sensor_data_db"
	query := fmt.Sprintf("SELECT 1 FROM pg_database WHERE datname = '%s'", databaseName)
	var exists int

	err = db.QueryRow(query).Scan(&exists)
	if err != nil && err != sql.ErrNoRows {
		db.Close()
		return nil, fmt.Errorf("failed to check for database existence: %w", err)
	}

	if exists == 0 {
		_, err = db.Exec(`CREATE DATABASE sensor_data_db;`)
		if err != nil {
			db.Close()
			return nil, fmt.Errorf("failed to create database 'sensor_data_db': %w", err)
		}
		fmt.Println("Database created successfully!")
	}
	db.Close()

	// 2. Connect to the specific 'sensor_data_db'
	targetDB, err := sql.Open("pgx", "postgres://postgres:postgres@localhost:5432/sensor_data_db?sslmode=disable")
	if err != nil {
		return nil, fmt.Errorf("failed to open target database: %w", err)
	}

	fmt.Println("Connected to Postgres")
	return &PostgresData{DB: targetDB}, nil
}

// Close closes the database connection
func (c *PostgresData) Close() {
	if c.DB != nil {
		c.DB.Close()
	}
}

func (c *PostgresData) RawPostgresData() {
	// Create the sensor_data table if it doesn't exist.
	var exists int
	err := c.DB.QueryRow(`SELECT 1 FROM pg_tables WHERE tablename = 'sensor_data'`).Scan(&exists)
	if err != nil && err != sql.ErrNoRows {
		fmt.Printf("Failed to check for table existence: %v\n", err)
		return
	}

	if exists == 0 {
		_, err = c.DB.Exec(`
		CREATE TABLE IF NOT EXISTS sensor_data (
			recorded BIGINT NOT NULL,
			location VARCHAR(255) NOT NULL,
			sensor VARCHAR(255) NOT NULL,
			measurement VARCHAR(255) NOT NULL,
			units VARCHAR(255) NOT NULL,
			value NUMERIC(10, 2) NOT NULL
		);`)
		if err != nil {
			fmt.Printf("failed to create sensor_data table: %v\n", err)
			return
		}
		fmt.Println("Table created successfully!")
	}

	// Unmarshal JSON string into SensorDataPostgres struct
	var sensorDataPostgres SensorDataPostgres
	if err := json.Unmarshal([]byte(getRandomSensorDataJsonString()), &sensorDataPostgres); err != nil {
		fmt.Printf("Failed to parse json: %v\n", err)
		return
	}
	fmt.Println("JSON string converted to SensorDataPostgres struct")

	// Prepare the SQL statement for inserting sensor data
	var query = `INSERT INTO sensor_data (recorded, location, sensor, measurement, units, value)
		VALUES ($1, $2, $3, $4, $5, $6)`

	// Execute the insert statement
	_, err = c.DB.Exec(query,
		sensorDataPostgres.Recorded,
		sensorDataPostgres.Location,
		sensorDataPostgres.Sensor,
		sensorDataPostgres.Measurement,
		sensorDataPostgres.Units,
		sensorDataPostgres.Value)
	if err != nil {
		fmt.Printf("Failed to insert data: %v\n", err)
		return
	}
	fmt.Println("SensorDataPostgres stored")

	// Query all rows from the sensor_data table
	var results []string
	rows, err := c.DB.Query(`SELECT recorded, location, sensor, measurement, units, value FROM sensor_data`)
	if err != nil {
		fmt.Printf("failed to query data: %v\n", err)
		return
	}
	defer rows.Close()

	for rows.Next() {
		var sensorDataPostgres SensorDataPostgres
		err := rows.Scan(
			&sensorDataPostgres.Recorded,
			&sensorDataPostgres.Location,
			&sensorDataPostgres.Sensor,
			&sensorDataPostgres.Measurement,
			&sensorDataPostgres.Units,
			&sensorDataPostgres.Value,
		)
		if err != nil {
			fmt.Printf("failed to scan row: %v\n", err)
			continue
		}

		jsonBytes, err := json.Marshal(sensorDataPostgres)
		if err != nil {
			fmt.Printf("failed to marshal row to JSON: %v\n", err)
			continue
		}
		results = append(results, string(jsonBytes))
	}

	if len(results) > 0 {
		data := []string{"[" + strings.Join(results, ",") + "]"}
		csvString, err := jsonToCSV(strings.Join(data, "\n"))
		if err != nil {
			fmt.Println(err)
		} else {
			fmt.Printf("%s\n%s", "SensorDataPostgres retrieved", csvString)
		}
	}

	// Delete all records
	_, err = c.DB.Exec("DELETE FROM sensor_data")
	if err != nil {
		fmt.Printf("Failed to delete data: %v\n", err)
		return
	}
	fmt.Println("SensorDataPostgres purged")
}
